crates/pep508-rs/src/marker.rs

```diff
@@ -899,195 +899,439 @@ impl<'a> TryFrom<MarkerEnvironmentBuilder<'a>> for MarkerEnvironment {
+    ///
+    /// Reports a warning on failure, and returns `None`.
+    fn version_inverted(
+        value: &str,
+        marker_operator: MarkerOperator,
+        key: MarkerValueVersion,
+        reporter: &mut impl Reporter,
+    ) -> Option<MarkerExpression> {
+        let version = match value.parse::<Version>() {
+            Ok(version) => version,
+            Err(err) => {
+                reporter.report(
+                    MarkerWarningKind::Pep440Error,
+                    format!(
+                        "Expected PEP 440 version to compare with {}, found {}, will evaluate to false: {}",
+                        key, value, err
+                    ),
+                );
+
+                return None;
+            }
+        };
+
+        let Some(operator) = marker_operator.to_pep440_operator() else {
+            reporter.report(
+                    MarkerWarningKind::Pep440Error,
+                    format!(
+                        "Expected PEP 440 version operator to compare {} with '{}', found '{}', will evaluate to false",
+                        key,
+                        version,
+                        marker_operator
+                    ),
+                );
+
+            return None;
+        };
+
+        Some(MarkerExpression::VersionInverted {
+            version,
+            operator,
+            key,
+        })
+    }
+
+    /// Creates an instance of `MarkerExpression::Extra` with the given values, falling back to
+    /// `MarkerExpression::Arbitrary` on failure.
+    fn extra(
+        operator: MarkerOperator,
+        value: &str,
+        reporter: &mut impl Reporter,
+    ) -> Option<MarkerExpression> {
+        let name = match ExtraName::from_str(value) {
+            Ok(name) => name,
+            Err(err) => {
+                reporter.report(
+                    MarkerWarningKind::ExtraInvalidComparison,
+                    format!("Expected extra name, found '{value}', will evaluate to false: {err}"),
+                );
+
+                return None;
+            }
+        };
+
+        match ExtraOperator::from_marker_operator(operator.clone()) {
+            Some(operator) => Some(MarkerExpression::Extra { operator, name }),
+            None => {
+                reporter.report(MarkerWarningKind::ExtraInvalidComparison, "Comparing extra with something other than a quoted string is wrong, will evaluate to false".to_string());
+                None
+            }
+        }
+    }
+
+    /// Evaluate a <`marker_value`> <`marker_op`> <`marker_value`> expression
+    ///
+    /// When `env` is `None`, all expressions that reference the environment
+    /// will evaluate as `true`.
+    fn evaluate(
+        &self,
+        env: Option<&MarkerEnvironment>,
+        extras: &[ExtraName],
+        reporter: &mut impl Reporter,
+    ) -> bool {
+        match self {
+            MarkerExpression::Version { key, specifier } => env
```

Comment:
The option thing feels odd (why are we evaluating markers against an environment when there is no environment?), can you do a `let Some(env) = env else { return false; }` or something like it?

Reply:
I added the `Option` in an older PR. If there's no environment, we specifically want all environment marker expressions to evaluate to `true`. But we still evaluate marker expressions involving extras.

This has the effect of evaluating a marker expression in an environment independent way. But since you still need to evaluate extras, you can't just do `return true` when `env` is `None` at the top here. But maybe it could be put at the top of some of the branches to simplify things a little.

It's an odd thing about PEP 508. Everything about marker expressions is tied to the environment _except_ for extras.

Reply:
This is a discussion beyond this PR, but i've been thinking about turning `MarkerEnvironment` from a dataclass into something more general, e.g. a trait where you can query a value or an expression in and the env can decides which fraction of the fields to evaluate. This would unify the cases for python version, markers and full environment behind a shared interface.

Reply:
I was thinking about something like that too, but the full shape of what it should look like wasn't totally clear to me. What was driving that thought is that I'm not a fan of spreading `Option<&MarkerEnvironment>` everywhere, and I would rather encapsulate that case analysis instead of pushing it out to everyone else.

So... yeah I'm on board with your idea. :-)